src/hotspot/share/interpreter/interpreterRuntime.cpp

```diff
@@ -733,20 +733,16 @@ JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorenter(JavaThread* thread, Ba
 JRT_END
 
 
-//%note monitor_1
-JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
+JRT_LEAF(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
 #ifdef ASSERT
   thread->last_frame().interpreter_frame_verify_monitor(elem);
 #endif
-  Handle h_obj(thread, elem->obj());
-  assert(Universe::heap()->is_in_or_null(h_obj()),
-         "must be NULL or an object");
-  if (elem == NULL || h_obj()->is_unlocked()) {
-    THROW(vmSymbols::java_lang_IllegalMonitorStateException());
-  }
-  ObjectSynchronizer::exit(h_obj(), elem->lock(), thread);
-  // Free entry. This must be done here, since a pending exception might be installed on
-  // exit. If it is not cleared, the exception handling code will try to unlock the monitor again.
+  oop obj = elem->obj();
+  assert(!obj->is_unlocked(), "caller checked these conditions");
+  assert(Universe::heap()->is_in_or_null(obj), "must be NULL or an object");
```

Comment:
obj can't be null at this point

Reply:
Agreed. If 'obj == NULL' on L740, the L741 would crash in ASSERT enabled bits.
The same is true of the original code if 'obj == NULL':
    old L744: if (elem == NULL || h_obj()->is_unlocked()) {
we would crash on is_unlocked() call.

Reply:
Yes should change assert to Universe::heap()->is_in(obj).